### PR TITLE
fix: Optimize AcceleratedParticle and MovingParticle

### PR DESCRIPTION
**Pull request by spydon**
# Description

Remove creation of new `Vector2` objects in `AcceleratedParticle` and `MovingParticle`.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples`.

## Breaking Change

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

## Related Issues

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

File: packages/flame/lib/src/particles/accelerated_particle.dart [hunk 1]
```diff
@@ -40,10 +40,27 @@ class AcceleratedParticle extends CurvedParticle with SingleChildParticle {
...
+
   @override
   void update(double dt) {
-    speed += acceleration * dt;
-    position += speed * dt - (acceleration * dt * dt) / 2;
+    speed.add(
+      _tmpVectors[0]
+        ..setFrom(acceleration)
+        ..scale(dt),
+    );
```
Comment (st-pasha) [inline packages/flame/lib/src/particles/accelerated_particle.dart:52]:
```suggestion
    speed.addScaled(acceleration, dt);
```

File: packages/flame/lib/src/particles/accelerated_particle.dart [hunk 2]
```diff
@@ -40,10 +40,27 @@ class AcceleratedParticle extends CurvedParticle with SingleChildParticle {
...
+      _tmpVectors[0]
+        ..setFrom(speed)
+        ..scale(dt)
+        ..sub(
+          _tmpVectors[1]
+            ..setFrom(acceleration)
+            ..scale(dt * dt)
+            ..scale(0.5),
+        ),
+    );
```
Comment (st-pasha) [inline packages/flame/lib/src/particles/accelerated_particle.dart:63]:
```suggestion
    position
      ..addScaled(speed, dt)
      ..addScaled(acceleration, -dt*dt*0.5);
```

File: packages/flame/lib/src/particles/moving_particle.dart [hunk 3]
```diff
@@ -24,10 +24,15 @@ class MovingParticle extends CurvedParticle with SingleChildParticle {
   })  : from = from ?? Vector2.zero(),
         super(lifespan: lifespan, curve: curve);
 
+  /// Used to avoid creating new [Vector2] objects in [update].
+  final _tmpVector = Vector2.zero();
```
Comment (st-pasha) [inline packages/flame/lib/src/particles/moving_particle.dart:28]:
```suggestion
  static final _tmpVector = Vector2.zero();
```

File: packages/flame/lib/src/particles/accelerated_particle.dart (hunk 1 above)
Comment (spydon) [reply packages/flame/lib/src/particles/accelerated_particle.dart]: Nice! I had forgotten about the existence of `addScaled`.

File: packages/flame/lib/src/particles/accelerated_particle.dart [hunk 4]
```diff
@@ -40,10 +40,16 @@ class AcceleratedParticle extends CurvedParticle with SingleChildParticle {
     canvas.restore();
   }
 
+  /// Used to avoid creating new vectors in [update].
+  static final _tmp1 = Vector2.zero();
+  static final _tmp2 = Vector2.zero();
+
```
Comment (spydon) [inline packages/flame/lib/src/particles/accelerated_particle.dart:46]:
```suggestion
```

File: packages/flame/lib/src/particles/accelerated_particle.dart [hunk 5]
```diff
@@ -40,10 +40,27 @@ class AcceleratedParticle extends CurvedParticle with SingleChildParticle {
     canvas.restore();
   }
 
+  /// Used to avoid creating new vectors in [update].
+  final _tmpVectors = [Vector2.zero(), Vector2.zero()];
```
[the remark this answers is not in the document]
Comment (spydon) [reply packages/flame/lib/src/particles/accelerated_particle.dart]: These could be removed completely instead.

**Review by wolfenrain — APPROVED** [review]
LGTM!

**Pull request merged**